invisible_cities/cities/beersheba.py

```diff
@@ -0,0 +1,355 @@
+                      deconv_mode     : str='joint',
+                      n_dim           : int=2,
+                      cut_type        : str='abs',
+                      inter_method    : str='cubic'):
+
+    """
+    Applies Lucy Richardson deconvolution to SiPM response with a
+    given set of PSFs and parameters.
+
+    Parameters
+    ----------
+    psf_fname       : Point-spread function.
+    e_cut            : Cut in relative value to the max voxel over the deconvolution output.
+    n_iterations    : Number of Lucy-Richardson iterations
+    n_iterations_g  : Number of Lucy-Richardson iterations for gaussian in 'separate mode'
+    iteration_tol   : Stopping threshold (difference between iterations).
+    sample_width    : Sampling size of the sensors.
+    bin_size        : Size of the interpolated bins.
+    energy_type     : Energy type ('E' or 'Ec', see Esmeralda) used for assignment.
+    deconv_mode     : 'joint' or 'separate', 1 or 2 step deconvolution, see description later.
+    diffusion       : Diffusion coefficients in each dimension for 'separate' mode.
+    n_dim           : Number of dimensions to apply the method (2 or 3).
+    cut_type        : Cut mode to the deconvolution output ('abs' or 'rel') using e_cut
+                      'abs': cut on the absolute value of the hits.
+                      'rel': cut on the relative value (to the max) of the hits.
+    inter_method    : Interpolation method.
+
+    Returns
+    ----------
+    apply_deconvolution : Function that takes hits and returns the
+    deconvolved data.
+    """
+    dimensions    = np.array(['X', 'Y', 'Z'][:n_dim])
+    sample_width  = np.array(sample_width   [:n_dim])
+    bin_size      = np.array(bin_size       [:n_dim])
+    diffusion     = np.array(diffusion      [:n_dim])
+
+    psfs          = pd.read_hdf(psf_fname)
+    deconvolution = deconvolve(n_iterations, iteration_tol, sample_width, bin_size, InterpolationMethod(inter_method))
+
+    e_type        = HitEnergy(energy_type)
+    cut_mode      = CutType(cut_type)
+    deco_mode     = DeconvolutionMode(deconv_mode)
+
+    def deconvolve_hits(df, z):
+        if   deco_mode is DeconvolutionMode.joint:
+            psf = psfs.loc[(psfs.z == find_nearest(psfs.z,                 z)) &
+                           (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                           (psfs.y == find_nearest(psfs.y, df.Ypeak.unique()))  , :]
+            deconv_image, pos = deconvolution(tuple(df.loc[:, v].values for v in dimensions), df.Q.values, psf)
+        elif deco_mode is DeconvolutionMode.separate:
+            psf_z0 = psfs.loc[(psfs.z == find_nearest(psfs.z,                 0)) &
+                              (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                              (psfs.y == find_nearest(psfs.y, df.Ypeak.unique()))  , :]
+            deconv_image, pos = deconvolution(tuple(df.loc[:, v].values for v in dimensions), df.Q.values, psf_z0)
```

Comment:
I don't think the `else` statement is need as the validity of the deco_mode is tested once you call the Enum object, right?

Aside from that, I'm ok with the rest.

Reply:
> the validity of the deco_mode is tested once you call the Enum object, right?

Not sure what you are you saying here. If you mean _"if it is not this value one, it must be the other one"_ then it is not the case. This is a common situation that leads to bugs. For instance, you may mistake this argument with another one and it wouldn't fail cause there is not an exhaustive check of the validity of the arguments. `Enum`s are used precisely to narrow down the allowed values of a given parameter to just a few options. If the user sends a wrong argument it should be detected.

Reply:
What I'm trying to say is that, when you create a DeconvolutionMode() object, if the argument passed is not one of the available modes, then it will fail and you won't even get to the inside of that function. 

The argument is a string which is converted at the beginning to a DeconvolutionMode object. The DeconvolutionMode object will fail (rightfully so) at the very beginning if the argument is not part of the `Enum` possibilites. So if you give a non-string argument to the city, it will fail because the DeconvolutionMode object won't swallow it, if it's a string, then it will fail if the string is not in the considered modes in the `Enum`. 

Or is there something else that I'm missing?